snuba/migrations/group_loader.py

```diff
@@ -0,0 +1,308 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from importlib import import_module
+from typing import Sequence
+
+from snuba.migrations.errors import MigrationDoesNotExist
+from snuba.migrations.migration import Migration
+
+
+class GroupLoader(ABC):
+    """
+    Provides the list of migrations associated with an group and a loader that returns
+    the requested migration.
+    """
+
+    @abstractmethod
+    def get_migrations(self) -> Sequence[str]:
+        """
+        Returns the list of migration IDs in the order they should be executed.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def load_migration(self, migration_id: str) -> Migration:
+        raise NotImplementedError
+
+
+class DirectoryLoader(GroupLoader, ABC):
+    """
+    Loads migrations that are defined as files of a directory. The file name
+    represents the migration ID.
+    """
+
+    def __init__(self, module: str) -> None:
+        self.__module = module
+
+    @abstractmethod
+    def get_migrations(self) -> Sequence[str]:
+        raise NotImplementedError
+
+    def load_migration(self, migration_id: str) -> Migration:
+        try:
+            module = import_module(f"{self.__module}.{migration_id}")
+            return module.Migration()  # type: ignore
+        except ModuleNotFoundError:
+            raise MigrationDoesNotExist("Invalid migration ID")
+
+
+class SystemLoader(DirectoryLoader):
+    def __init__(self) -> None:
+        super().__init__("snuba.migrations.system_migrations")
+
+    def get_migrations(self) -> Sequence[str]:
+        return ["0001_migrations"]
+
+
+class EventsLoader(DirectoryLoader):
+    def __init__(self) -> None:
+        super().__init__("snuba.snuba_migrations.events")
+
+    def get_migrations(self) -> Sequence[str]:
```

Comment:
suggestion: what if we just used the `import_modules_in_subdirectory` function that we use for loading registered classes in order to do this and avoid the extra step of having to populate these lists?

Reply:
Seems like that would make sense. Looking at how `get_migrations() -> Sequence[str]` works, I think the refactor would actually be a big change. We don't want to completely remove this string list because the index in this list is used to track reverse migrations.